File: Mao/160.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        nodea = headA
        nodeb = headB

        while nodea != nodeb:
		        nodea = headB if not nodea else nodea.next
	        	nodeb = headA if not nodeb else nodeb.next
        return nodea

File: shadow/160.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        ListNode p1 = headA;
        ListNode p2 = headB;

        while (p1 != null || p2 != null) {
            if (p1 == p2) {
                return p1;
            }

            if (p1 == null) {
                p1 = headB;
            }
            else {
                p1 = p1.next;
            }

            if (p2 == null) {
                p2 = headA;
            }
            else {
                p2 = p2.next;
            }
        }

        return null;
    }
}
